### PR TITLE
nixos/influxdb2: automatic initial setup and nixos tests

### PR DESCRIPTION
**Pull request by oddlama**
## Description of changes

This PR adds provisioning options and nixos tests to influxdb2. This allows skipping the initial setup page.

- Provisioning was split out into second PR: #249502

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
@ofborg test influxdb2

**Comment by NickCao** [comment]
Disclaimer: I'm no longer using influxdb.

Overall I like the idea of declarative setup, but the implementation looks overly complicated. Can we first get the part that can be done with `influx setup` in and figure out better ways to do the rest?

**Comment by oddlama** [comment]
> Overall I like the idea of declarative setup, but the implementation looks overly complicated. Can we first get the part that can be done with `influx setup` in and figure out better ways to do the rest?

Sure we could do that, but how would you propose to make this simpler? Since there's no first class provisioning in influx there'll always be a lot of "manual" work to get it done

**Comment by NickCao** [comment]
> Sure we could do that, but how would you propose to make this simpler? Since there's no first class provisioning in influx there'll always be a lot of "manual" work to get it done

That's why I said "figure out", maybe ansible, maybe terraform, surely there can be better ways than the current scripting.

**Comment by oddlama** [comment]
> surely there can be better ways than the current scripting

I'm happy for any suggestions. Although I just couldn't come with anything simpler that doesn't involve building a influxdb client from scratch..

Anyhow, this has now been reduced to the initial setup part, the provisioning support is now tracked in #249502 (which will have to be rebased if/when this is merged). Let's continue talking over there.

File: nixos/modules/services/databases/influxdb2.nix [hunk 1]
```diff
@@ -53,6 +110,57 @@ in
...
+      postStart = let
+        initCfg = cfg.provision.initialSetup;
+      in mkIf cfg.provision.enable (
+        ''
+          set -euo pipefail
+          export INFLUX_HOST="http://"${escapeShellArg (cfg.settings.http-bind-address or "localhost:8086")}
+
+          # Wait for the influxdb server to come online
+          count=0
+          while ! ${getExe pkgs.influxdb2-cli} ping &>/dev/null; do
```
Comment (NickCao) [inline nixos/modules/services/databases/influxdb2.nix:130]: We can add `influxdb2-cli` to the `PATH` of the service by setting `systemd.services.<name>.path`

File: nixos/modules/services/databases/influxdb2.nix [hunk 2]
```diff
@@ -53,6 +110,57 @@ in
...
+            if [ "$count" -eq 300 ]; then
+              echo "Tried for 30 seconds, giving up..."
+              exit 1
+            fi
+
+            if ! kill -0 "$MAINPID"; then
+              echo "Main server died, giving up..."
+              exit 1
+            fi
```
Comment (NickCao) [inline nixos/modules/services/databases/influxdb2.nix:139]: When this happens, wouldn't systemd also kill postStart?

File: nixos/modules/services/databases/influxdb2.nix [hunk 3]
```diff
@@ -53,6 +110,57 @@ in
...
+          # Do the initial database setup. Pass /dev/null as configs-path to
+          # avoid saving the token as the active config.
+          if test -e "$STATE_DIRECTORY/.first_startup"; then
+            ${getExe pkgs.influxdb2-cli} setup \
+              --configs-path /dev/null \
+              --org ${escapeShellArg initCfg.organization} \
+              --bucket ${escapeShellArg initCfg.bucket} \
+              --username ${escapeShellArg initCfg.username} \
+              --password "$(< ${escapeShellArg initCfg.passwordFile})" \
+              --token "$(< ${escapeShellArg initCfg.tokenFile})" \
```
Comment (NickCao) [inline nixos/modules/services/databases/influxdb2.nix:154]: What would happen when `tokenFile` is `null`

File: nixos/modules/services/databases/influxdb2.nix [hunk 4]
```diff
@@ -53,6 +110,57 @@ in
...
+              --retention ${escapeShellArg initCfg.retention} \
+              --force >/dev/null
+
+            rm -f "$STATE_DIRECTORY/.first_startup"
+          fi
+
+          export INFLUX_TOKEN=$(< ${escapeShellArg initCfg.tokenFile})
```
Comment (NickCao) [inline nixos/modules/services/databases/influxdb2.nix:161]: Why this?

File: nixos/tests/influxdb2.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      tokenArg = "--token verysecureadmintoken";
+    in ''
+      machine.wait_for_unit("influxdb2.service")
+
+      out = machine.succeed("influx org list ${tokenArg}")
+      assert "default" in out
+
+      out = machine.succeed("influx bucket list ${tokenArg} --org default")
+      assert "default" in out
+    '';
```
[the remark this answers is not in the document]
Comment (NickCao) [reply nixos/tests/influxdb2.nix:32]: (Or if the password is only useful with the webui? We can skip this if that's the case)

File: nixos/modules/services/databases/influxdb2.nix (hunk 2 above)
Comment (oddlama) [reply nixos/modules/services/databases/influxdb2.nix:139]: No, from [systemd.service](https://www.freedesktop.org/software/systemd/man/systemd.service.html):
> Note that if any of the commands specified in ExecStartPre=, ExecStart=, or ExecStartPost= fail (and are not prefixed with "-", see above) or time out before the service is fully up, execution continues with commands specified in ExecStopPost=, the commands in ExecStop= are skipped.

File: nixos/modules/services/databases/influxdb2.nix (hunk 3 above)
Comment (oddlama) [reply nixos/modules/services/databases/influxdb2.nix:154]: Good point, either we can make it non optional or skip specifying it when it is null. The problem with the latter is that full provisioning requires this token to be known, so maybe making it mandatory is the right call. What are your thoughts?

File: nixos/modules/services/databases/influxdb2.nix (hunk 4 above)
Comment (oddlama) [reply nixos/modules/services/databases/influxdb2.nix:161]: Originally that was part of the full provisioning, just forgot to remove it.

File: nixos/modules/services/databases/influxdb2.nix (hunk 3 above)
Comment (NickCao) [reply nixos/modules/services/databases/influxdb2.nix:154]: We can default it to `$STATE_DIRECTORY/token` and generate it if non-exist. But for the sake of simplicity and the declarative spirit of nixos, let's just make it mandatory until someone complains.

Comment (NickCao) [reply nixos/modules/services/databases/influxdb2.nix:154]: Also we can (and should IMO) make use of `LoadCredential`.

Comment (oddlama) [reply nixos/modules/services/databases/influxdb2.nix]: Done

File: nixos/modules/services/databases/influxdb2.nix [hunk 6]
```diff
@@ -52,7 +107,59 @@ in
...
+          # Do the initial database setup. Pass /dev/null as configs-path to
+          # avoid saving the token as the active config.
+          if test -e "$STATE_DIRECTORY/.first_startup"; then
+            influx setup \
+              --configs-path /dev/null \
+              --org ${escapeShellArg initCfg.organization} \
+              --bucket ${escapeShellArg initCfg.bucket} \
+              --username ${escapeShellArg initCfg.username} \
+              --password "$(< ${escapeShellArg initCfg.passwordFile})" \
+              --token "$(< "$CREDENTIALS_DIRECTORY/admin-token")" \
```
Comment (NickCao) [inline nixos/modules/services/databases/influxdb2.nix:155]: Same for `passwordFile`

File: nixos/tests/influxdb2.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (oddlama) [reply nixos/tests/influxdb2.nix:35]: Technically only for the UI, but the session token is technically a fully fledged api token with most permissions. I've added as short curl invocation to make sure it's set correctly.

File: nixos/modules/services/databases/influxdb2.nix (hunk 6 above)
Comment (oddlama) [reply nixos/modules/services/databases/influxdb2.nix:158]: Obviously, whoops.

**Review by NickCao — APPROVED** [review]
diff LGTM, waiting for ofborg.

**Comment by chvp** [comment]
The `passwordFile` and `tokenFile` options are used even if provisioning is not enabled. Is this intentional?

**Comment by NickCao** [comment]
> The `passwordFile` and `tokenFile` options are used even if provisioning is not enabled. Is this intentional?

Oops that's not.

**Comment by oddlama** [comment]
Definitely not intended, i thought I guarded against that here

https://github.com/NixOS/nixpkgs/blob/8a294d53a8d940ccd4b8a567f8cbfb35827061bc/nixos/modules/services/databases/influxdb2.nix#L127

**Pull request merged**